aten/src/ATen/native/cuda/LinearAlgebra.cu

```diff
@@ -23,4 +24,128 @@ Tensor& bmm_out_cuda(Tensor &result, const Tensor& batch1, const Tensor& batch2)
+{
+#if CUDA_VERSION < 9200 && CUDA_VERSION >= 9000
+  static std::once_flag alreadyWarned;
+  const int LIMIT = 1 << 21;
+  if (i_m > LIMIT || i_n > LIMIT || i_k > LIMIT) {
+    cudaDeviceProp* prop = at::cuda::getCurrentDeviceProperties();
+    if (prop->major == 5 || prop->major == 6) {
+      std::call_once(alreadyWarned, []() {
+        TORCH_WARN("Matrix multiplication for dimensions larger than 2^21 has known bugs on your combination of CUDA version and device type. Please consider upgrading to CUDA 9.2 or later.");
+      });
+    }
+  }
+#endif
+}
+
+Tensor& addmm_out_cuda_impl(Tensor& result, const Tensor& self, const Tensor& mat1, const Tensor& mat2, Scalar beta, Scalar alpha) {
+  TORCH_CHECK(
+    (mat1.dim() == 2) && (mat2.dim() == 2) &&
+    (self.dim() == 2) && (result.dim() == 2),
+    "tensors must be 2-D"
+  );
+  IntArrayRef mat1_sizes = mat1.sizes();
+  IntArrayRef mat2_sizes = mat2.sizes();
+  IntArrayRef self_sizes = self.sizes();
+  TORCH_CHECK(mat1_sizes[1] == mat2_sizes[0], "mat1 dim 1 must match mat2 dim 0");
+  TORCH_CHECK(self_sizes[0] == mat1_sizes[0], "self dim 0 must match mat1 dim 0");
+  TORCH_CHECK(self_sizes[1] == mat2_sizes[1], "self dim 1 must match mat2 dim 1");
+
+  // If self and result either point to the same data or if beta is zero,
+  // we can avoid copying self into result. Otherwise, we need to copy.
+  if (result.data_ptr() != self.data_ptr()) {
+    if (beta.to<double>() != 0.0) {
+      result.copy_(self);
+    }
+  }
+
+  IntArrayRef result_sizes = result.sizes();
+  if ((result_sizes[0] == 0) || (result_sizes[1] == 0)) {
+    return result;
+  }
+
+  bool transpose_result;
+  Tensor result_ = prepare_matrix_for_cublas(result, transpose_result);
+  bool transpose_mat1;
+  bool transpose_mat2;
+  Tensor mat1_ = transpose_result ? mat2 : mat1;
+  Tensor mat2_ = transpose_result ? mat1 : mat2;
+  mat1_ = prepare_matrix_for_cublas(mat1_, transpose_mat1);
+  mat2_ = prepare_matrix_for_cublas(mat2_, transpose_mat2);
+
+  if (transpose_result) {
+    transpose_mat1 = !transpose_mat1;
+    transpose_mat2 = !transpose_mat2;
+    mat1_sizes = mat1_.sizes();
+    mat2_sizes = mat2_.sizes();
+  }
+
+  int64_t m = mat1_sizes[transpose_result ? 1 : 0];
+  int64_t k = mat1_sizes[transpose_result ? 0 : 1];
+  int64_t n = mat2_sizes[transpose_result ? 0 : 1];
+  int64_t mat1_ld = mat1_.stride((transpose_mat1 == transpose_result) ? 1 : 0);
+  int64_t mat2_ld = mat2_.stride((transpose_mat2 == transpose_result) ? 1 : 0);
+  int64_t result_ld = result_.stride(transpose_result ? 0 : 1);
+  at::ScalarType scalar_type = self.scalar_type();
+
+  AT_DISPATCH_FLOATING_TYPES_AND2(at::ScalarType::Half, at::ScalarType::BFloat16, scalar_type, "addmm_cuda", [&] {
+    if (scalar_type == at::ScalarType::BFloat16 || scalar_type == at::ScalarType::Int) {
+      checkCuda90Bug(static_cast<int>(m), static_cast<int>(n), static_cast<int>(k));
+    }
+    scalar_t alpha_val = alpha.to<scalar_t>();
```

Comment:
I'm generally unhappy that for half type alpha and beta will be cast to half, and possibly overflow/underflow, just to be cast back to float again, but it's not introduced by this PR and semi-hardcoded in at::cuda::blas::gemm, so not your fault. In 99.999% of cases they will be 0/1 anyway.